### PR TITLE
Fail mount when the block size changes

### PR DESCRIPTION
**Pull request by colin-foster-in-advantage**
When the on-disk block size doesn't match the config block size, it is
possible to get file corruption. For instance, if the num blocks was
0x200 and we re-mount with 0x100 files could be corrupt.

If we re-mount with a larger number of blocks things should be safer,
but could be handled with a resize option or perhaps a mount flag to
ignore this parameter.

### CODE REVIEW

**Comment by preetpalkang** [comment]
The change looks good, but we should also check the `block_size` in addition to `block_count`

We should merge this and get a new release :)

**Comment by colin-foster-in-advantage** [comment]
> The change looks good, but we should also check the `block_size` in addition to `block_count`

Done. In hindsight I wasn't clear about size vs count. Count is the one that is probably more likely to change, whereas size is most likely dependent on the hardware.

**Review by preetpalkang — APPROVED** [review]
Sorry for small scope creep as the #584 was only meant to handle changes to `block_count`.

Regardless, I think it is a good idea because we should have sanity checks to ensure that the `lfs_config` differs from the `superblock` and the read/write/erase functions are not going to work, hence we want to force the user to reformat.

This is a great change until something like #279 is implemented. I'd recommend merging to the master.

File: lfs.c [hunk 1]
```diff
@@ -3761,6 +3761,13 @@ static int lfs_rawmount(lfs_t *lfs, const struct lfs_config *cfg) {
 
                 lfs->attr_max = superblock.attr_max;
             }
+
+            if ((superblock.block_count != lfs->cfg->block_count) ||
+                (superblock.block_size != lfs->cfg->block_size))
+            {
```
Comment (preetpalkang) [inline lfs.c:3767]: nit picking: brace should be on previous line for consistency of code base.

Comment (colin-foster-in-advantage) [reply lfs.c]: Agreed. Fixed below. I'm not sure how well github works with squashing commits. Let me know if there's anything I should do for that.

Comment (preetpalkang) [reply lfs.c]: looks good!

@maintainer let's approve and merge this :)

**Comment by geky** [comment]
This looks reasonable to me, thanks for creating this PR. Sorry about the delay on my end, doing a bit of catch-up.

File: lfs.c [hunk 2]
```diff
@@ -3761,6 +3761,12 @@ static int lfs_rawmount(lfs_t *lfs, const struct lfs_config *cfg) {
 
                 lfs->attr_max = superblock.attr_max;
             }
+
+            if ((superblock.block_count != lfs->cfg->block_count) ||
+                (superblock.block_size != lfs->cfg->block_size)) {
+                err = LFS_ERR_CORRUPT;
```
[the remark this answers is not in the document]
Comment (colin-foster-in-advantage) [reply lfs.c]: Thanks for the feedback. I made these changes and pushed them to my branch. I can squash the commits if you prefer that.

**Comment by geky** [comment]
Thanks for updating this based on the feedback, looks good and will bring this in on the next minor release.

I agree with @dpgeorge and @preetpalkang, this is a good protection to have, at least until https://github.com/littlefs-project/littlefs/issues/279 is resolved (though a shrinking filesystem will probably always be a hard error).

---

> whereas size is most likely dependent on the hardware.

Quick side note, but the `block_size` can actually be a multiple of the physical erase size, which you might do for performance reasons. So this should be treated roughly the same as the `block_count`, but determining both at mount time has some problems  (https://github.com/littlefs-project/littlefs/issues/349)

**Review by geky — APPROVED** [review]

**Comment by geky** [comment]
Thanks for putting together this PR!

**Pull request merged**